File: mqttapi/app/api/routes/environment.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel

from app.api.deps import get_current_user, get_db
from app.db import Database

router = APIRouter(prefix="/api/v1", tags=["environment"])


@router.get("/environment/devices")
def list_environment_devices(
    db: Database = Depends(get_db),
    _user: dict = Depends(get_current_user),
) -> list[dict[str, Any]]:
    """WingOnIOT environment device list (Environment_Device) with bound cell and room."""
    return db.list_environment_devices()


class BindDeviceCellRequest(BaseModel):
    floor_id: int
    row_no: int
    col_no: int


@router.post("/environment/devices/{sn}/cell")
def bind_device_cell(
    sn: str,
    body: BindDeviceCellRequest,
    db: Database = Depends(get_db),
    _user: dict = Depends(get_current_user),
) -> dict[str, Any]:
    """Bind the device to a specific cell (device → cell; replaces the previous binding)."""
    result = db.bind_device_cell(sn, body.floor_id, body.row_no, body.col_no)
    if result != "ok":
        if result == "floor_mismatch":
            raise HTTPException(status_code=409, detail="Floor mismatch: device floor differs from cell floor")
        raise HTTPException(status_code=404, detail="Device or cell not found")
    return {"ok": True}


@router.delete("/environment/devices/{sn}/cell")
def unbind_device_cell(
    sn: str,
    db: Database = Depends(get_db),
    _user: dict = Depends(get_current_user),
) -> dict[str, Any]:
    """Unbind all cell bindings of the device."""
    ok = db.unbind_device_cell(sn)
    if not ok:
        raise HTTPException(status_code=404, detail="No binding to remove")
    return {"ok": True}


@router.get("/environment/monitoring")
def list_environment_monitoring(
    limit: int = Query(default=50, ge=1, le=1000, description="Max rows"),
    offset: int = Query(default=0, ge=0),
    db: Database = Depends(get_db),
    _user: dict = Depends(get_current_user),
) -> dict[str, Any]:
    """Paginated Environmental_Monitoring list (newest writes first)."""
    items, total = db.list_environment_monitoring(limit=limit, offset=offset)
    return {"total": total, "limit": limit, "offset": offset, "items": items}


@router.get("/environment/floor-summary")
def floor_environment_summary(
    db: Database = Depends(get_db),
    _user: dict = Depends(get_current_user),
) -> list[dict[str, Any]]:
    """Latest temperature/humidity aggregated per floor (median of each device's latest record)."""
    return db.floor_environment_summary()

File: frontend/src/api/environment.ts
import api from './http'

/** 设备绑定的格子（building_cell）坐标，x/z 为前端 3D 平面世界坐标 */
export interface DeviceCell {
  cell_id: number
  row_no: number
  col_no: number
  x: number
  z: number
}

export interface EnvironmentDevice {
  sn: string
  name: string | null
  deviceName: string | null
  model: string | null
  floor: string | null
  /** 映射到 3D 楼栋的层号（B2/F→1、B1/F→2、4/F→4…），无法解析为 null */
  level: number | null
  location: string | null
  macAddress: string | null
  /** 最新一条监测记录（无数据为 null） */
  toDateTime: string | null
  temperatureMedian: number | null
  humidityMedian: number | null
  /** 绑定的格子（null = 未绑定，含大厅设备） */
  cell: DeviceCell | null
  /** 格子所属房间业务键 room_id（null = 大厅/走廊格子） */
  room_id: string | null
}

export interface EnvironmentReading {
  id: number
  sn: string | null
  fromDateTime: string | null
  toDateTime: string | null
  temperatureMin: number | null
  temperatureMedian: number | null
  temperatureMax: number | null
  humidityMin: number | null
  humidityMedian: number | null
  humidityMax: number | null
  InsertAt: string | null
}

export interface FloorEnvironmentSummary {
  /** WingOnIOT 原始楼层字符串，如 '4/F'、'B1/F' */
  floor: string | null
  /** 映射到 3D 楼栋的层号（B2/F→1、B1/F→2、4/F→4…），无法解析为 null */
  level: number | null
  temperature: number | null
  humidity: number | null
  device_count: number
  updated_at: string | null
}

export function listEnvironmentDevices() {
  return api.get<EnvironmentDevice[]>('/api/v1/environment/devices')
}

export function listEnvironmentReadings(params?: {
  limit?: number
  offset?: number
}) {
  return api.get<{ total: number; limit: number; offset: number; items: EnvironmentReading[] }>(
    '/api/v1/environment/monitoring',
    { params },
  )
}

export function getFloorEnvironmentSummary() {
  return api.get<FloorEnvironmentSummary[]>('/api/v1/environment/floor-summary')
}

/** 绑定设备到具体格子（设备→格子；替换设备原有绑定） */
export function bindDeviceToCell(
  sn: string,
  params: { floor_id: number; row_no: number; col_no: number },
) {
  return api.post<{ ok: boolean }>(`/api/v1/environment/devices/${sn}/cell`, params)
}

/** 解绑设备的所有格子绑定 */
export function unbindDeviceFromCell(sn: string) {
  return api.delete<{ ok: boolean }>(`/api/v1/environment/devices/${sn}/cell`)
}
